### ISSUE TITLE
DataTable with dynamic number of rows is unstable and breaks

### ISSUE THREAD

**Issue by pmicheletty** [bokeh/bokeh#7035]
DataTable with a dynamic number of rows is very unstable and breaks in bokeh 0.12.7 and 0.12.9. This bug was reproducible with data_table.py example from github: https://github.com/bokeh/bokeh/blob/master/examples/models/server/data_tables.py

See screenshot below for example. If you start the app and change the selections in the dropdowns (more changes- more unstable) eventually it will break, specifically if the updates change the number of rows in the table. The ColumnDataSource continues to update correctly (as seen in the plot) but the table doesn't.

We have found a slight hack that if you force the number of rows in the table to always be the same, it doesn't have this issue.

![image](https://user-images.githubusercontent.com/9287000/31395822-f0368c9e-ad9e-11e7-8585-7a840bb9c4c2.png)

Tested on Windows 10, with Google Chrome browser.
Current python environment:
```
#
alabaster                 0.7.10                   py36_0  
astroid                   1.5.3                    py36_0  
babel                     2.5.0                    py36_0  
bleach                    1.5.0                    py36_0  
bokeh                     0.12.9                     py_0    bokeh
bzip2                     1.0.6                    vc14_3  [vc14]
certifi                   2016.2.28                py36_0  
chardet                   3.0.4                    py36_0  
colorama                  0.3.9                    py36_0  
cycler                    0.10.0                   py36_0  
decorator                 4.1.2                    py36_0  
docutils                  0.14                     py36_0  
entrypoints               0.2.3                    py36_0  
futures-compat            1.0                       py3_0  
h5py                      2.7.0               np112py36_0  
hdf5                      1.8.15.1                 vc14_4  [vc14]
html5lib                  0.9999999                py36_0  
icu                       57.1                     vc14_0  [vc14]
imagesize                 0.7.1                    py36_0  
ipykernel                 4.6.1                    py36_0  
ipyparallel               6.0.2                    py36_0  
ipython                   6.1.0                    py36_0  
ipython_genutils          0.2.0                    py36_0  
isort                     4.2.15                   py36_0  
jedi                      0.10.2                   py36_2  
jinja2                    2.9.6                    py36_0  
jpeg                      9b                       vc14_0  [vc14]
jsonschema                2.6.0                    py36_0  
jupyter_client            5.1.0                    py36_0  
jupyter_core              4.3.0                    py36_0  
lazy-object-proxy         1.3.1                    py36_0  
libpng                    1.6.30                   vc14_1  [vc14]
markupsafe                1.0                      py36_0  
matplotlib                2.0.2               np112py36_0  
mistune                   0.7.4                    py36_0  
mkl                       2017.0.3                      0  
nbconvert                 5.2.1                    py36_0  
nbformat                  4.4.0                    py36_0  
numexpr                   2.6.2               np112py36_0  
numpy                     1.12.1                   py36_0  
numpydoc                  0.7.0                    py36_0  
openssl                   1.0.2l                   vc14_0  [vc14]
pandas                    0.19.2              np112py36_1  
pandocfilters             1.4.2                    py36_0  
path.py                   10.3.1                   py36_0  
pickleshare               0.7.4                    py36_0  
pip                       9.0.1                    py36_1  
prompt_toolkit            1.0.15                   py36_0  
psutil                    5.2.2                    py36_0  
pycodestyle               2.3.1                    py36_0  
pyflakes                  1.6.0                    py36_0  
pygments                  2.2.0                    py36_0  
pylint                    1.7.2                    py36_0  
pyparsing                 2.2.0                    py36_0  
pyqt                      5.6.0                    py36_2  
pyshp                     1.2.12                    <pip>
pytables                  3.2.2               np112py36_4  
python                    3.6.2                         0  
python-dateutil           2.6.1                    py36_0  
pytz                      2017.2                   py36_0  
pyyaml                    3.12                     py36_0  
pyzmq                     16.0.2                   py36_0  
qt                        5.6.2                    vc14_6  [vc14]
qtawesome                 0.4.4                    py36_0  
qtconsole                 4.3.1                    py36_0  
qtpy                      1.3.1                    py36_0  
requests                  2.14.2                   py36_0  
rope                      0.9.4                    py36_1  
scipy                     0.19.1              np112py36_0  
setuptools                36.4.0                   py36_0  
simplegeneric             0.8.1                    py36_1  
singledispatch            3.4.0.3                  py36_0  
sip                       4.18                     py36_0  
six                       1.10.0                   py36_0  
snowballstemmer           1.2.1                    py36_0  
sphinx                    1.6.3                    py36_0  
sphinxcontrib             1.0                      py36_0  
sphinxcontrib-websupport  1.0.1                    py36_0  
spyder                    3.2.3                    py36_0  
testpath                  0.3.1                    py36_0  
tk                        8.5.18                   vc14_0  [vc14]
tornado                   4.5.2                    py36_0  
traitlets                 4.3.2                    py36_0  
vs2015_runtime            14.0.25420                    0  
wcwidth                   0.1.7                    py36_0  
wheel                     0.29.0                   py36_0  
wincertstore              0.2                      py36_0  
wrapt                     1.10.11                  py36_0  
xlrd                      1.1.0                    py36_0  
zlib                      1.2.11                   vc14_0  [vc14]
```

**Comment by bryevdv** [comment]
Thanks @pmicheletty I had noticed this and meant to submit an issue but got distracted before I did.

**Comment by bryevdv** [comment]
@clairetang6 I believe this is a `CDSView issue`. When this `DataTableView` method is called:
```
  updateGrid: () ->
    @data.constructor(@model.source, @model.view)
    @grid.invalidate()
    @grid.render()
```
the `model.source` has the correct updated data, but the `model.view.indices` value is the original. I.e. in this example selecting a make results in a smaller CDS with say 37 entries, but `view.indices` is still of length 237 or whatever. Is this an update ordering issue?

**Comment by bryevdv** [comment]
As a point of reference, this diff "fixes" but is certainly not the correct solution. In particular I expect it would fall over if the CDS data columns go *bigger*
```
diff --git a/bokehjs/src/coffee/models/widgets/tables/data_table.coffee b/bokehjs/src/coffee/models/widgets/tables/data_table.coffee
index dc680b137..8020a2778 100644
--- a/bokehjs/src/coffee/models/widgets/tables/data_table.coffee
+++ b/bokehjs/src/coffee/models/widgets/tables/data_table.coffee
@@ -19,7 +19,7 @@ export class DataProvider
   constructor: (@source, @view) ->
     if DTINDEX_NAME of @source.data
       throw new Error("special name #{DTINDEX_NAME} cannot be used as a data table column")
-    @index = @view.indices
+    @index = @view.indices.slice(0, @source.get_length())

   getLength: () -> @index.length

```

**Comment by bryevdv** [comment]
I can also state that this call to `updateGrid` was triggered by an emit on `this.model.source.properties.data.change`. What/when is the CDSView update triggered? wondering if e.g. it needs to happen first before anything else but is not. (Just a thought)

**Comment by clairetang6** [comment]
The CDSView is supposed to update its indices when the data source changes. I see that in the CDSView there is a signal connection `@connect(@source.change, () -> @compute_indices())`. I'm still not 100% clear on when the `source.change` is emitted versus `source.properties.data.change`, but I imagine that adding another signal connection in the CDSView for the latter would fix this issue.

**Comment by clairetang6** [comment]
You also wouldn't want to just slice a view's indices to the correct length, because the view indices takes filtering into account. The indices should be calculated through the view's `compute_indices` method which takes filters and length into account.

**Comment by bryevdv** [comment]
> You also wouldn't want to just slice a view's indices to the correct length, because the view indices takes filtering into account

 I know it's not a solution, I was only trying to confirm and illustrate the source of the issue. I will try adding the extra connection.

**Comment by bryevdv** [comment]
@clairetang6 This change does *not* fix:
```
diff --git a/bokehjs/src/coffee/models/sources/cds_view.coffee b/bokehjs/src/coffee/models/sources/cds_view.coffee
index 06ca3cd6d..1c9432b85 100644
--- a/bokehjs/src/coffee/models/sources/cds_view.coffee
+++ b/bokehjs/src/coffee/models/sources/cds_view.coffee
@@ -28,6 +28,8 @@ export class CDSView extends Model
       @compute_indices()
       @change.emit()
     )
+    if @source?properties?.data?.change?
+      @connect(@source.properties.data.change, () -> @compute_indices())
     if @source?.change?
       @connect(@source.change, () -> @compute_indices())
     if @source?.streaming?
```
This change *does*, at least "fix". I would rate this as much closer to "correct" but still probably not ideal unless there is no other option.
```
diff --git a/bokehjs/src/coffee/models/widgets/tables/data_table.coffee b/bokehjs/src/coffee/models/widgets/tables/data_table.coffee
index dc680b137..b12eb5a5d 100644
--- a/bokehjs/src/coffee/models/widgets/tables/data_table.coffee
+++ b/bokehjs/src/coffee/models/widgets/tables/data_table.coffee
@@ -19,6 +19,7 @@ export class DataProvider
   constructor: (@source, @view) ->
     if DTINDEX_NAME of @source.data
       throw new Error("special name #{DTINDEX_NAME} cannot be used as a data table column")
+    @view.compute_indices()
     @index = @view.indices

   getLength: () -> @index.length
```

**Comment by clairetang6** [comment]
Hmm, it could be that in the first case the view's indices are not being updated until after `updateGrid` is already called.. I think that may have been what you were talking about earlier in terms of the order of the updates. The second diff above seems like it will always produce the correct behavior. It just would be lead to extra calls to `compute_indices`. 

The ideal solution would be some way of guaranteeing that `compute_indices` happens before `updateGrid`. I'm not suggesting we do this (without further consideration), but one thought is that computing the indices is something that has to the do with the "models" while updating the grid is something to do with "views". Maybe changes should update "model"-related things before "view"-related things. (I'm overloading these words a little, but I mean abstract properties in the models versus things to do with rendering. Also, just some initial thoughts and will need to think about this more).

**Comment by bryevdv** [comment]
So my thoughts pretty much echo yours. I guess a compromise would be to merge this fix for now and make a separate issue to explore imposing ordering constraints or priorities on signal callbacks. Let me know if you think that is reasonable.

**Comment by clairetang6** [comment]
I think that sounds reasonable. 

Another issue this reminds me of is `map_data` being called on on every render (if I remember that correctly), where it really only needs to be called when certain properties have changed, but calling it every time does work.

**Comment by bryevdv** [comment]
Yes exactly. OK I will work up a small PR. I do want to try and find tests to add.
